### ISSUE TITLE
Monitoring ExternalWork time

### ISSUE THREAD

**Issue by kpedro88** [cms-sw/cmssw#29142]
#29141 introduces SONIC to CMSSW, including a client class that monitors how much time is spent in ExternalWork calls. The client works as follows:
* `SonicEDProducer::acquire()` calls `SonicClient::predict()` and returns
* `SonicClient::predict()` stores the start time `t0`, does its work, then calls `SonicClient::finish()`
* `SonicClient::finish()` checks the time again (`t1`) and emits a `LogInfo` message with the time difference `t1-t0`

It would be useful to generalize this feature to all ExternalWork calls. This is possibly already tracked in the stall monitor, but for users, a summary from the TimerService or similar entity would be more accessible.

Ideally, we could track two signals:
* `acquire()` returns
* `produce()` added to the task queue
However, I am not sure what would be needed to implement this.

**Comment by kpedro88** [comment]
assign core

**Comment by makortel** [comment]
Just to clarify, this would be about measuring the time between
* `acquire()` returns
* `WaitingTaskWithArenaHolder::doneWaiting()` is called

? The former already has a signal (`PostModuleEventAcquire`), so the question becomes how to deal with the latter. To think out loud some further thoughts
* How to deal with `WaitingTaskWithArenaHolder::makeWaitingTaskHolderAndRelease()`
* The question is not really about `WaitingTaskWithArenaHolder`, but how it is used in conjunction with `ExternalWork` (although that is currently the only use for `WaitingTaskWithArenaHolder`)
* Should the functionality be limited to time measurement only, or would a specific ActivityRegistry signal be better? Either way the work done there should be minimal given that the code would run on non-TBB thread.
* I'd imagine with a functionality like `edm::async()` (#29188) it would be straightforward to add a notification point that would not have to be dealt explicitly by user, but that would imply that all uses of `ExternalWork` would have to use such facility for the synchronization. On the other hand, there doesn't seem to be that many offloading(-like) APIs that support callback functions.

**Comment by kpedro88** [comment]
Yes, that is what I think we should measure.

I suppose a general ActivityRegistry signal could be useful in some cases - I don't know how all of our timing/profiling services work under the hood, but some of them might need it.

Can we just make `doneWaiting()` issue the signal whenever it's called? I guess the problem then would be double-counting if ExternalWork were used in a blocking mode. Could add an argument to `doneWaiting()` so it can be called to suppress the signal...

I'm not entirely sure I understand how `edm::async()` would integrate into ExternalWork universally. Can you elaborate? (Also, thanks for the reminder of that issue, had forgotten about it.)

**Comment by makortel** [comment]
> Can we just make `doneWaiting()` issue the signal whenever it's called?

No, the waiting system should stay independent of `ActivityRegistry` (its abstraction is on lower level).

> I'm not entirely sure I understand how `edm::async()` would integrate into ExternalWork universally. Can you elaborate?

`edm::async()` would be a function taking `edm::WaitingTaskWithArenaHolder` and a functor arguments. The functor would be run on a separate, non-TBB thread, and it would call the blocking synchronization on a given API, but not deal with `WaitingTaskWithArenaHolder` explicitly. The internals of `edm::async()` would deal with `WaitingTaskArenaHolder`, and thus in principle would have a point where e.g. a signal could be emitted. I haven't thought this approach thoroughly (so I'm not sure if this is a good idea or not).

**Comment by kpedro88** [comment]
So in SONIC terms, you're basically proposing to use PseudoAsync for everything, even true Async. If this doesn't add substantial overhead, it could be feasible. My question, though, is how do we "enforce" (ideally, automate) this usage - would the ExternalWork interface change?

**Comment by makortel** [comment]
Well, I'm not really proposing anything yet, just thinking out loud. Enforcement is clearly a weak point here.

By the way, do you know how Triton client implements the asynchronous call? Does it itself have some thread(s) waiting around?

**Comment by kpedro88** [comment]
Yes, Triton uses an `std::thread` for the asynchronous call: https://github.com/triton-inference-server/client/blob/cd3bc6859d150b6e07408537779f35242dac9117/src/c%2B%2B/library/grpc_client.cc#L947-L960
